contracts/multisig-prover/src/contract.rs

```diff
@@ -1,44 +1,479 @@
+            block
+                .height
+                .try_into()
+                .expect("violated invariant: block height cannot be invalid"),
+            config.signing_threshold,
+            participants,
+        ))
     }
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn query(_deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+pub fn reply(deps: DepsMut, _env: Env, reply: Reply) -> Result<Response, ContractError> {
+    match parse_reply_execute_data(reply) {
+        Ok(MsgExecuteContractResponse { data: Some(data) }) => {
+            let command_batch_id = REPLY_BATCH.load(deps.storage)?;
+
+            let session_id =
+                from_binary(&data).map_err(|_| ContractError::InvalidContractReply {
+                    reason: "invalid multisig session ID".to_string(),
+                })?;
+
+            let proof_id = ProofID::new(&command_batch_id, &session_id);
+
+            PROOF_BATCH_MULTISIG.save(deps.storage, &proof_id, &(command_batch_id, session_id))?;
+
+            Ok(Response::new().add_event(Event::ProofUnderConstruction { proof_id }.into()))
+        }
+        Ok(MsgExecuteContractResponse { data: None }) => Err(ContractError::InvalidContractReply {
+            reason: "no data".to_string(),
+        }),
+        Err(_) => {
+            unreachable!("violated invariant: replied failed submessage with ReplyOn::Success")
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
     match msg {
-        QueryMsg::GetProof { proof_id } => to_binary(&query::get_proof(proof_id)?),
+        QueryMsg::GetProof { proof_id } => to_binary(&query::get_proof(deps, proof_id)?),
     }
 }
 
 pub mod query {
     use super::*;
 
-    pub fn get_proof(_proof_id: String) -> StdResult<GetProofResponse> {
-        todo!()
+    pub fn get_proof(deps: Deps, proof_id: String) -> StdResult<GetProofResponse> {
+        let config = CONFIG.load(deps.storage)?;
+
+        let proof_id = HexBinary::from_hex(proof_id.as_str())?.into();
+        let (batch_id, session_id) = PROOF_BATCH_MULTISIG.load(deps.storage, &proof_id)?;
+
+        let batch = COMMANDS_BATCH.load(deps.storage, &batch_id)?;
+
+        let query_msg = multisig::msg::QueryMsg::GetMultisig { session_id };
+
+        let multisig: Multisig = deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: config.multisig.to_string(),
+            msg: to_binary(&query_msg)?,
+        }))?;
+
+        let status = match multisig.state {
+            MultisigState::Pending => ProofStatus::Pending,
+            MultisigState::Completed => {
+                let execute_data = batch
+                    .encode_execute_data(multisig.quorum, multisig.signers)
+                    .map_err(|e| {
+                        StdError::generic_err(format!("failed to encode execute data: {}", e))
+                    })?;
```

Comment:
in general, don't create hierarchical errors manually. That's exactly what error_stack is trying to solve. Create an error for the specific inner reason, then change the context to the more general error. Also, try to use generic errors only to convert external errors and then go with typed errors from there

Reply:
moved to https://axelarnetwork.atlassian.net/browse/AXE-1613